File: java/src/main/java/com/experian/dvs/client/address/search/ResultFuture.java
package com.experian.dvs.client.address.search;

import com.experian.dvs.client.exceptions.EDVSException;
import com.experian.dvs.client.server.RestApiResponseError;
import com.experian.dvs.client.server.address.search.RestApiAddressSearchResponse;

import java.util.Optional;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Future;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

public class ResultFuture implements Future<Result> {

    private final Future<RestApiAddressSearchResponse> apiFuture;

    public ResultFuture(Future<RestApiAddressSearchResponse> apiFuture) {
        this.apiFuture = apiFuture;
    }

    @Override
    public boolean cancel(boolean mayInterruptIfRunning) {
        return this.apiFuture.cancel(mayInterruptIfRunning);
    }

    @Override
    public boolean isCancelled() {
        return this.apiFuture.isCancelled();
    }

    @Override
    public boolean isDone() {
        return this.apiFuture.isDone();
    }

    @Override
    public Result get() throws InterruptedException, ExecutionException {
        final RestApiAddressSearchResponse apiSearchResponse = this.apiFuture.get();
        final Optional<RestApiResponseError> optError = apiSearchResponse.getError();
        if (optError.isPresent()) {
          throw EDVSException.using(optError.get());
        }
        return new Result(apiSearchResponse);
    }

    @Override
    public Result get(long timeout, TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException {
        final RestApiAddressSearchResponse apiSearchResponse = this.apiFuture.get(timeout, unit);
        final Optional<RestApiResponseError> optError = apiSearchResponse.getError();
        if (optError.isPresent()) {
            throw EDVSException.using(optError.get());
        }
        return new Result(apiSearchResponse);
    }
}

File: csharp/DVSClient/DVSClient/address/search/ResultFuture.cs
using DVSClient.Exceptions;
using DVSClient.Server.Address.Search;

namespace DVSClient.Address.Search
{
    public class ResultFuture
    {
        private readonly Task<RestApiAddressSearchResponse> _apiFuture;

        public ResultFuture(Task<RestApiAddressSearchResponse> apiFuture)
        {
            _apiFuture = apiFuture;
        }

        public async Task<Result> GetAsync()
        {
            try
            {
                var apiSearchResponse = await _apiFuture;
                var optError = apiSearchResponse.Error;
                if (optError != null)
                {
                    throw EDVSException.Using(optError);
                }
                return new Result(apiSearchResponse);
            }
            catch (Exception e) when (e is TaskCanceledException || e is OperationCanceledException)
            {
                throw new RestApiInterruptionOrExecutionException(e);
            }
        }

        public async Task<Result> GetAsync(TimeSpan timeout)
        {
            try
            {
                var apiSearchResponse = await Task.WhenAny(_apiFuture, Task.Delay(timeout)) == _apiFuture
                    ? await _apiFuture
                    : throw new TimeoutException();
                var optError = apiSearchResponse.Error;
                if (optError != null)
                {
                    throw EDVSException.Using(optError);
                }
                return new Result(apiSearchResponse);
            }
            catch (Exception e) when (e is TaskCanceledException || e is OperationCanceledException || e is TimeoutException)
            {
                throw new RestApiInterruptionOrExecutionException(e);
            }
        }

        public bool Cancel(bool mayInterruptIfRunning)
        {
            return _apiFuture.IsCanceled;
        }

        public bool IsCancelled()
        {
            return _apiFuture.IsCanceled;
        }

        public bool IsDone()
        {
            return _apiFuture.IsCompleted;
        }
    }
}